docs/guides/2sl2b-design.md

```diff
@@ -0,0 +1,96 @@
+---
+slug: /2sl2b/design
+displayed_sidebar: "0.3"
+---
+
+# Design
+
+## Goals
+
+NOTE: this currently focuses more on new goals in cloak (v0.3) relative to europa (v0.2); it does not yet include much about other general Dagger goals that have been retained between versions (like portability).
+
+### API-centric
+
+Functionality in Dagger should be exposed through language-agnostic APIs. This includes both:
+
+- "Core" APIs, which operate as Dagger's "standard library" and provide basic primitives like the ability to pull container images, execute commands, etc.
+- Extensions, which are like external libraries that build higher-level tools and abstractions on top of the primitives provided by the core APIs and other extensions.
+
+The API framework Dagger uses should enable:
+
+1. Calling of automation from anywhere and any language
+   - The API may be invoked from a CLI command, from a script written in any of our supported languages, from an HTTP request sent by a browser, etc.
+1. Authoring of automation in any language
+   - The caller of an API doesn't need to know what language the API is implemented with.
+   - This freedom leaves both caller and callee to choose whichever language fits their particular job the best without having to think about interoperability.
+   - Rather than reimplementing the same automation in every possible programming language, it only needs to be implemented once and s then re-usable by all.
+   - Strengths of different languages thus enhance each other rather than creating silos.
+     - If Python has the world's greatest library for accomplishing X, an extension author should feel free to import+use it; they will be making that functionality available for anyone else in Go, Typescript, Bash and any other Dagger-supported language along the way.
```

Comment:
Does that mean you can call the library outside Dagger container?
For instance directly call kubectl Go SDK or AWS Python SDK instead of CLI?

Reply:
Yes, you can see some example that do some work on the host directly with some Go package, and my Go extension was called. It was itself doing some Go code directly.
But it could have been called by a python code loading some env with pure python code and sending that to the Go extension.

Reply:
Go code calling the extension: https://github.com/dagger/cloak/pull/191/files#diff-a0005cdb0414f59fb56c24e008b5abdbc5167b1ff6d0683fd078fe020da0e81dR46
The extension called: https://github.com/dagger/cloak/pull/191/files#diff-de000dd72fc1ae924965a9ca3cc9eddf762864de4f3c78dab61ad7a62f8e4248R20
But they could be different languages altogether.
Also, it is an old code. Much has changed since.

Reply:
That's impressive and really crazy :O 
But then how do you manage os compatibilities? If you execute operations on host, there's no buildkit nor qemu involved so how will it be managed? Imagine a Go extension that only work on Armv8 for some reason, what will happens on Windows?

Reply:
It's a chance than in general Go is pretty portable :p But yeah, it could be an issue.